// nodez/nodez.js
function s4() {
  return Math.floor((1 + Math.random()) * 0x10000)
             .toString(16)
             .substring(1);
}
function guid() {
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
         s4() + '-' + s4() + s4() + s4();
}

exports["GeneralNode"] = function (label,nm,data) {
	this.label = label
	this.uuid = guid()
	this.revision = 1
	this.creationDate = new Date().toISOString()
	this.lastModifiedDate = new Date().toISOString()
    this.name = nm
    for (d in data) {
    	this[d]=data[d]
    }
	this.toNeo4j = function () {
		var str=""
		for (i in this) {
			switch (Object.prototype.toString.call(this[i])) {
				case "[object String]": {
					str=str+" "+i+" "+this[i]+" "+" "+Object.prototype.toString.call(this[i])+"\n"
					break
				}
				case "[object Number]": {
					str=str+" "+i+" "+this[i]+" "+" "+Object.prototype.toString.call(this[i])+"\n"
					break
				}
			}
		}
		return str
	}
	this.ExtractNeo4jData = function () {
		var data = {}
		for (i in this) {
			switch (Object.prototype.toString.call(this[i])) {
				case "[object String]": {
					data[i]=this[i]
 					break
				}
				case "[object Number]": {
					data[i]=this[i]
					break
				}
				case "[object Array]": {
					data[i]=this[i]
					break
				}
			}
		}
		return data
	}
	this.save = function(d,_CB_) {
		var DBROUTE="http://localhost:7474/db/data/"
		options={ 
			url:DBROUTE+'node',
			method:"POST",
			headers:{'content-type':'application/json'},
			body:JSON.stringify(d.ExtractNeo4jData())
			}
		console.log("Calling Neo4j broker with:"+JSON.stringify(options))
		var request = require('request');
		request(options,afterPosting)
	}
	this.cypherSave = function(d,_CB_) {
		var DBROUTE="http://localhost:7474/db/data/cypher/"
//		cypher=""
//		cypher=cypher+'{ 'query' : 'create (n:"+d.ExtractNeo4jData().label+" { props } ) return n ',"
//		cypher=cypher+"'params': { 'props' :  "+JSON.stringify(d.ExtractNeo4jData())+" } }"
		cypher={
			"query":"create (n:"+d.ExtractNeo4jData().label+" { props } ) return n",
			"params": {
				"props": d.ExtractNeo4jData()
			}
		}
		console.log(JSON.stringify(cypher))
		options={
			url:DBROUTE,
			method:"POST",
			headers:{'content-type':'application/json'},
			body:JSON.stringify(cypher)
			}
		console.log("Calling Neo4j broker with:"+options)
		var request = require('request');
		request(options,_CB_)
	}	
}
